polygeist/lib/Dialect/Polygeist/Analysis/MemoryAccessAnalysis.cpp

```diff
@@ -90,36 +90,243 @@ static bool isSmallerThanNegativeOne(Value val, DataFlowSolver &solver) {
+
+  Value getValue() const { return val; }
+
+  /// Determine whether the multiplier has value one.
+  bool isOne(DataFlowSolver &solver) const {
+    if (auto constVal = dyn_cast<arith::ConstantIndexOp>(val.getDefiningOp())) {
+      if (constVal.value() == 1)
+        return true;
+    }
+    std::optional<APInt> constVal = getConstIntegerValue(val, solver);
+    return (constVal.has_value() && constVal->isOne());
+  }
+
+  /// Create a multiplier with value one.
+  static Multiplier one(MLIRContext *ctx) {
+    OpBuilder b(ctx);
+    return Multiplier(b.create<arith::ConstantIndexOp>(b.getUnknownLoc(), 1));
+  };
+
+private:
+  Value val;
+};
+
+raw_ostream &operator<<(raw_ostream &os, const Multiplier &multiplier) {
+  return os << multiplier.val;
+}
+
+/// Represents a multiplier or a value.
+class ValueOrMultiplier {
+  friend raw_ostream &operator<<(raw_ostream &, const ValueOrMultiplier &);
+
+public:
+  ValueOrMultiplier(Multiplier mul) : data(mul) {}
+  ValueOrMultiplier(Value val) : data(val) {}
+
+  bool isMultiplier() const { return std::holds_alternative<Multiplier>(data); }
+  bool isValue() const { return std::holds_alternative<Value>(data); }
+
+  Multiplier getMultiplier() const {
+    assert(isMultiplier() && "Expecting multiplier");
+    return std::get<Multiplier>(data);
+  }
+
+  Value getValue() const {
+    assert(isValue() && "Expecting value");
+    return std::get<Value>(data);
+  }
+
+  /// Determine whether this class holds the value one.
+  bool isOne(DataFlowSolver &solver) const {
+    if (isMultiplier())
+      return getMultiplier().isOne(solver);
+
+    Value val = getValue();
+    if (auto constVal = dyn_cast<arith::ConstantIndexOp>(val.getDefiningOp())) {
+      if (constVal.value() == 1)
+        return true;
+    }
+    std::optional<APInt> constVal = getConstIntegerValue(val, solver);
+    return (constVal.has_value() && constVal->isOne());
   }
 
-  Value other;
+private:
+  std::variant<Value, Multiplier> data;
+};
+
+[[maybe_unused]] raw_ostream &
+operator<<(raw_ostream &os, const ValueOrMultiplier &valOrMultiplier) {
+  if (valOrMultiplier.isMultiplier())
+    return os << valOrMultiplier.getMultiplier();
+  return os << valOrMultiplier.getValue();
+}
+
+} // namespace
+
+// Visit a binary operation of type \tparam T. The LHS and RHS operand of the
+// binary operation are processed by applying the function \p getMultiplier. The
+// result of the visit is computed via the \p computeResult function.
+template <typename T,
+          typename = std::enable_if_t<llvm::is_one_of<
+              T, arith::AddIOp, arith::SubIOp, arith::MulIOp>::value>>
+static ValueOrMultiplier visitBinaryOp(
+    T binOp, const Value factor, DataFlowSolver &solver,
+    std::function<ValueOrMultiplier(const Value, const Value, DataFlowSolver &)>
+        getMultiplier,
+    std::function<ValueOrMultiplier(ValueOrMultiplier, ValueOrMultiplier)>
+        computeResult) {
```

Comment:
Can we have template arguments instead of `std::function` for the input function argument types?

Reply:
done